src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,630 @@
+		return;
+	}
+	ts_WaitForBackgroundWorkerStartup(tsbgw_he->db_scheduler_handle, &worker_pid);
+	tsbgw_message_send_ack(message, true);
+	return;
+}
+
+
+/*
+ * Drain queue, handle each message synchronously.
+ */
+static void
+launcher_handle_messages(HTAB *db_htab)
+{
+	TsbgwMessage *message;
+
+	for (message = tsbgw_message_receive(); message != NULL; message = tsbgw_message_receive())
+	{
+
+		PGPROC	   *sender;
+		VirtualTransactionId vxid;
+
+		sender = BackendPidGetProc(message->sender_pid);
+		if (sender != NULL)
+			GET_VXID_FROM_PGPROC(vxid, *sender);
+		else
+		{
+			ereport(LOG, (errmsg("timescalebgw cluster launcher received message from non-existent backend")));
+			continue;
+		}
+
+		switch (message->message_type)
+		{
+			case START:
+				message_start_action(db_htab, message, vxid);
+				break;
+			case STOP:
+				message_stop_action(db_htab, message);
+				break;
+			case RESTART:
+				message_restart_action(db_htab, message, vxid);
+				break;
+		}
+
+	}
+}
+
+
+static void
+tsbgw_sigterm(SIGNAL_ARGS)
+{
+	ereport(LOG, (errmsg("timescaledb launcher terminated by administrator command. Launcher will not restart after exiting.")));
+	proc_exit(0);
+}
+
+static void
+tsbgw_sigint(SIGNAL_ARGS)
+{
+	ereport(ERROR, (errmsg("timescaledb launcher canceled by administrator command. Launcher will restart after exiting")));
+}
+
+extern void
+tsbgw_cluster_launcher_main(void)
+{
+
+	int			n;
+	HTAB	   *db_htab;
+	int			num_unreserved_workers;
+
+	pqsignal(SIGTERM, tsbgw_sigterm);
+	pqsignal(SIGINT, tsbgw_sigint);
+	BackgroundWorkerUnblockSignals();
+
+	if (!tsbgw_total_workers_increment())
+	{
+		/*
+		 * should be the first thing happening so if we already exceeded our
+		 * limits it means we have a limit of 0 and we should just exit
+		 */
+		ereport(LOG, (errmsg("timescale background worker limit set to %d. please set higher if you would like to use background workers.", guc_max_bgw_processes)));
+		proc_exit(0);
+	}
+	/* Connect to the db, no db name yet, so can only access shared catalogs */
+	BackgroundWorkerInitializeConnection(NULL, NULL);
+	pgstat_report_appname("Timescale BGW Cluster Launcher");
+	ereport(LOG, (errmsg("timesacle bgw cluster launcher started")));
+
+	tsbgw_message_queue_set_reader();
+	db_htab = populate_database_htab();
+	before_shmem_exit(launcher_pre_shmem_cleanup, (Datum) db_htab);
+
+	num_unreserved_workers = guc_max_bgw_processes - tsbgw_total_workers_get();
+	if (num_unreserved_workers < hash_get_num_entries(db_htab))
+		ereport(LOG, (errmsg("total databases = %ld timescale background worker limit set to %d. ", hash_get_num_entries(db_htab), guc_max_bgw_processes),
+					  errhint("you may start background workers manually by using the  _timescaledb_internal.start_background_workers() function in each database you would like to have a scheduler worker in.")));
+	else
+	{
+		for (n = 1; n <= hash_get_num_entries(db_htab); n++)
+			tsbgw_total_workers_increment();
```

Comment:
I haven't addressed this one, simply because it seemed like an optimization that wasn't horribly pressing, happy to address in a further pr if you feel it's necessary.